File: SolidSamples_Part2_OCP/app/src/main/java/com/example/solidsamples/good/java/AreaCalculator.java
package com.example.solidsamples.good.java;

/**
 * Created by pieter.snyman on 11/13/17.
 */

public class AreaCalculator {

    public double area(Shape[] shapes)
    {
        double area = 0;
        for (Shape shape : shapes) {
            area += shape.getArea();
        }

        return area;
    }
}

File: SolidSamples_Part3_LSP/app/src/main/java/com/example/solidsamples/good/kotlin/AreaCalculator.kt
package com.example.solidsamples.good.kotlin

/**
 * Created by pieter.snyman on 11/13/17.
 */
class AreaVerifier {

    private fun tester() {
        verify(Rectangle(5.0, 4.0))
        //Can't break this rule anymore
        verify(Square(5.0))
    }

    private fun verify(rectangle: Rectangle): Boolean {
        rectangle.width = 5.0
        rectangle.height = 4.0

        if (rectangle.area != 20.0) {
            throw Exception("5 x 4 = 20 expected for Rectangle")
        }

        return true
    }

    private fun verify(square: Square): Boolean {
        square.size = 5.0

        if (square.area != 25.0) {
            throw Exception("5 x 5 = 25 expected for Square")
        }

        return true
    }
}